### PR TITLE
Implement executable linking in LLVM (IR and AOT).

### PR DESCRIPTION
**Pull request by ScottTodd**
Fixes https://github.com/google/iree/issues/3137, follows https://github.com/google/iree/pull/3283.

This also refactors the LLVM targets to share more code, since they have the same behavior until serialization.

### CODE REVIEW

File: iree/compiler/Dialect/HAL/IR/HALOps.cpp [hunk 1]
```diff
@@ -1500,6 +1500,24 @@ ArrayAttr InterfaceOp::getExecutableSetLayoutsAttr() {
       })));
 }
 
+bool InterfaceOp::isEquivalentTo(InterfaceOp other) {
+  auto bindings = getBlock().getOps<InterfaceBindingOp>();
+  auto otherBindings = other.getBlock().getOps<InterfaceBindingOp>();
+  auto it = bindings.begin(), lhsEnd = bindings.end();
```
Comment (benvanik) [inline iree/compiler/Dialect/HAL/IR/HALOps.cpp:1506]: I'm sure there's some ADT helper for this; like
```
auto bindings = llvm::to_vector<4>(...);
auto otherBindings = llvm::to_vector<4>(...);
return bindings.size() == otherBindings.size() &&
       llvm::all_of(llvm::zip(bindings, otherBindings), [](InterfaceBindingOp lhs, InterfaceBindingOp rhs) {
         return OperationEquivalence::isEquvalentTo(lhs, rhs);
       });
```
(@silvasean may know a cleaner way)

File: iree/compiler/Dialect/HAL/IR/HALOps.td [hunk 2]
```diff
@@ -2124,6 +2124,8 @@ def HAL_InterfaceOp : HAL_Op<"interface", [
     // TODO(benvanik): replace with a nested typed attr that works.
     // Array of HAL_DescriptorSetLayoutBindingArrayAttr.
     ArrayAttr getExecutableSetLayoutsAttr();
+
+    bool isEquivalentTo(IREE::HAL::InterfaceOp other);
```
Comment (benvanik) [inline iree/compiler/Dialect/HAL/IR/HALOps.td:2128]:
```suggestion
    // Returns true if the all bindings in the interface match exactly those
    // in |other| (including order).
    bool isEquivalentTo(IREE::HAL::InterfaceOp other);
```

**Review by benvanik — APPROVED** [review]

File: iree/compiler/Dialect/HAL/IR/HALOps.cpp (hunk 1 above)
Comment (ScottTodd) [reply iree/compiler/Dialect/HAL/IR/HALOps.cpp:1506]: I was probably looking at https://github.com/llvm/llvm-project/blob/a67d164a82bc535025bb65736ee4e5f381500ed8/mlir/lib/Transforms/Utils/RegionUtils.cpp#L484-L497 when I wrote this originally.

Comment (ScottTodd) [reply iree/compiler/Dialect/HAL/IR/HALOps.cpp]: Ok, found similar code here using the structure you suggested: https://github.com/llvm/llvm-project/blob/f402e682d0ef5598eeffc9a21a691b03e602ff58/mlir/lib/Dialect/Traits.cpp#L21-L24 and updated to it.

**Pull request merged**